plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -73,14 +73,37 @@ type predicateMetadata struct {
 	podRequest                *schedulercache.Resource
 	podPorts                  map[int]bool
 	matchingAntiAffinityTerms []matchingPodAntiAffinityTerm
+
+	// CheckServiceAffinity: pods cache, otherwise this is a large query, will slow things down
+	// when there are 100s of services/pods.
+	lock        sync.Mutex
+	servicePods map[string][]*api.Pod
 }
 
 type matchingPodAntiAffinityTerm struct {
 	term *api.PodAffinityTerm
 	node *api.Node
 }
 
-func PredicateMetadata(pod *api.Pod, nodeInfoMap map[string]*schedulercache.NodeInfo) interface{} {
+// A set of functions which take in the predicateMetadata type, and add information to it.
+// This map should only be modified when registering predicates.
+var predicatePrecomputations map[string]func(pm *predicateMetadata) = make(map[string]func(pm *predicateMetadata))
+
+func RegisterPredicatePrecomputation(predicateName string, precomp func(pm *predicateMetadata)) {
```

Comment:
Please add a comment that this function is not thread-safe (or make it thread-safe?)